src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscovery.java

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.msal4j;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.util.Arrays;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+
+class AadInstanceDiscovery {
+    private static final Logger log = LoggerFactory.getLogger(AadInstanceDiscovery.class);
+
+    private final static TreeSet<String> TRUSTED_HOSTS_SET = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+
+    static {
+        TRUSTED_HOSTS_SET.addAll(Arrays.asList(
+                "login.windows.net",
+                "login.chinacloudapi.cn",
+                "login-us.microsoftonline.com",
+                "login.microsoftonline.de",
+                "login.microsoftonline.com",
+                "login.microsoftonline.us"));
+    }
+
+    private final static String DEFAULT_TRUSTED_HOST = "login.microsoftonline.com";
+
+    private final static String AUTHORIZE_ENDPOINT_TEMPLATE = "https://{host}/{tenant}/oauth2/v2.0/authorize";
+    private final static String INSTANCE_DISCOVERY_ENDPOINT_TEMPLATE = "https://{host}/common/discovery/instance";
+    private final static String INSTANCE_DISCOVERY_REQUEST_PARAMETERS_TEMPLATE =
+            "?api-version=1.1&authorization_endpoint={authorizeEndpoint}";
+
+    static ConcurrentHashMap<String, InstanceDiscoveryMetadataEntry> cache = new ConcurrentHashMap<>();
```

Comment:
Please be aware that as part of reliability work we are likely going to start persisting this to disk or enable hydrating it from the app